<a href="https://colab.research.google.com/github/adilsonsimiaoDK/StatisticToDS/blob/main/sampling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import random
import numpy as np
from sklearn.model_selection import StratifiedShuffleSplit


In [ ]:
dataset = pd.read_csv('census.csv')

# New Section

In [ ]:
dataset.shape

(32561, 15)

Simple Random Sampling: 
Simple randon sampling you have select value of sampling that you define to work in my case I decided n = 100 value random, the function(sample of library pandas) return 100 values random,  if you want to work always with the same values config random_state=1.

In [ ]:
dataset.tail()

age      workclass  final-weight  ... hour-per-week  native-country  income
32556   27        Private        257302  ...            38   United-States   <=50K
32557   40        Private        154374  ...            40   United-States    >50K
32558   58        Private        151910  ...            40   United-States   <=50K
32559   22        Private        201490  ...            20   United-States   <=50K
32560   52   Self-emp-inc        287927  ...            40   United-States    >50K

[5 rows x 15 columns]

In [ ]:
df_srs = dataset.sample(n=100, random_state=1)


In [ ]:
df_srs.head(10)

age          workclass  ...  native-country  income
9646    62   Self-emp-not-inc  ...   United-States   <=50K
709     18            Private  ...   United-States   <=50K
7385    25            Private  ...   United-States    >50K
16671   33            Private  ...   United-States   <=50K
21932   36            Private  ...   United-States   <=50K
31503   34            Private  ...   United-States    >50K
3308    43            Private  ...   United-States   <=50K
4208    47          Local-gov  ...   United-States    >50K
21078   29            Private  ...   United-States   <=50K
27354   23            Private  ...   United-States   <=50K

[10 rows x 15 columns]

In [ ]:
def simple_random_sampling(dataset, sampling):
  return dataset.sample(n=sampling) 

In [ ]:
df_srs = simple_random_sampling(dataset, 100)
df_srs.shape

(100, 15)

Systematic Sampling: 

1.   You have to get total rows with function shape of your dataset and divedid by number of sampling that you wan to work eg: my dataset has rows 32561, columns 15, you must get rows dataset/sampling or 32561//100 = 325 it will return a number int.
2. Yo must choose a value random 0 up to dataset in 325  



In [ ]:

def systematicsampling(dataset, sampling):
  interval = len(dataset)//sampling
  random.seed(1)  
  valueF = random.randint(0, interval)
  indices = np.arange(valueF, len(dataset), step=interval)
  systematic_sampling = dataset.iloc[indices]
  return systematic_sampling



In [ ]:
print(systematicsampling(df_srs,100))

       age          workclass  ...  native-country  income
28368   73            Private  ...   United-States   <=50K
2066    46            Private  ...   United-States    >50K
11611   36            Private  ...   United-States   <=50K
19711   24          Local-gov  ...           Haiti   <=50K
16973   26            Private  ...   United-States   <=50K
...    ...                ...  ...             ...     ...
10495   29            Private  ...   United-States   <=50K
4408    52            Private  ...   United-States   <=50K
20389   50   Self-emp-not-inc  ...   United-States   <=50K
18469   17            Private  ...   United-States   <=50K
6669    31            Private  ...   United-States    >50K

[100 rows x 15 columns]


Cluster for Sampling:
1. You have divide the dataset by number de sampling after that will be created groups that you should identify for id, 
2. command used: np.unique(group, return_counts=True) return the values not duplicate and count of element in each group.


In [ ]:
def cluster_sampling(dataset, number_group):
  group = []
  group_id = 0
  count = 0
  interval = len(dataset)/number_group
  for _ in dataset.iterrows():
      group.append(group_id)
      count += 1 
      if count > interval:
        count = 0
        group_id +=1
  dataset['group'] = group
  selected_group = random.randint(0, number_group)
  return dataset[dataset['group']== selected_group]
  


In [ ]:
group_sampling = cluster_sampling(dataset, 10)
group_sampling.shape, group_sampling.group.value_counts()
print(group_sampling)

       age      workclass  final-weight  ...  native-country  income group
29313   45        Private        285060  ...   United-States    >50K     9
29314   28      State-gov        189765  ...   United-States   <=50K     9
29315   23        Private        130905  ...   United-States   <=50K     9
29316   50        Private        146325  ...   United-States    >50K     9
29317   33        Private        102821  ...   United-States   <=50K     9
...    ...            ...           ...  ...             ...     ...   ...
32556   27        Private        257302  ...   United-States   <=50K     9
32557   40        Private        154374  ...   United-States    >50K     9
32558   58        Private        151910  ...   United-States   <=50K     9
32559   22        Private        201490  ...   United-States   <=50K     9
32560   52   Self-emp-inc        287927  ...   United-States    >50K     9

[3248 rows x 16 columns]


**Stratified sampling**
1. Sampling/variable = Stratified sampling

In [ ]:
dataset.income.value_counts()

 <=50K    24720
 >50K      7841
Name: income, dtype: int64

In [ ]:
7841 / len(dataset), 24720 / len(dataset)

(0.2408095574460244, 0.7591904425539756)

In [ ]:
split  = StratifiedShuffleSplit(test_size=0.0030711587481956942)
for x,y in split.split(dataset, dataset['income']):
  df_x =dataset.iloc[x]
  df_y =dataset.iloc[y]

In [ ]:
df_x.shape, df_y.shape

((32461, 16), (100, 16))

In [ ]:
sampling = 100  #size my sample
sampling / len(dataset)

0.0030711587481956942

In [ ]:
df_y['income'].value_counts()

 <=50K    76
 >50K     24
Name: income, dtype: int64